### PR TITLE
Core: Allow deleting old partition spec columns in V1

### LINKED ISSUE #7386: Issue /w UpdatePartitionSpec after deleting a partition field and schema field
### Apache Iceberg version

1.2.1 (latest release)

### Query engine

None

### Please describe the bug 🐞

With a table /w v1 metadata, after deleting a partition field and then deleting the related schema field, partitioning is then broken moving forward (cannot add a new partition field).

Steps:

1. Create a table (v1) with two columns in it (id 1, and 2 respectively) (col1 long ,col2 long) also /w partition spec col1 (identity)
2. Remove partition field from partition spec aka void(1)
3. Remove col1 from table
4. try to add a new partition field for col2 and receive exception 

```
java.lang.NullPointerException: Cannot invoke "String.isEmpty()" because "name" is null
	at org.apache.iceberg.Schema.findField(Schema.java:341)
	at org.apache.iceberg.TableMetadata.freshSpec(TableMetadata.java:743)
	at org.apache.iceberg.TableMetadata.access$2900(TableMetadata.java:50)
	at org.apache.iceberg.TableMetadata$Builder.addPartitionSpecInternal(TableMetadata.java:1462)
	at org.apache.iceberg.TableMetadata$Builder.setDefaultPartitionSpec(TableMetadata.java:1040)
	at org.apache.iceberg.TableMetadata.updatePartitionSpec(TableMetadata.java:552)
	at org.apache.iceberg.BaseUpdatePartitionSpec.commit(BaseUpdatePartitionSpec.java:331)
```

### PR DESCRIPTION
**Pull request by Fokko**
Resolves #7386

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 1]
```diff
@@ -386,4 +386,26 @@ public void testGroupingKeyTypeWithIncompatibleSpecEvolution() {
...
+  @Test
+  public void testDeletingPartitionField() {
+    TestTables.TestTable table =
+        TestTables.create(tableDir, "test", SCHEMA, BY_DATA_SPEC, V1_FORMAT_VERSION);
+
+    table.updateSpec().removeField("data").commit();
+
+    table.updateSchema().deleteColumn("data").commit();
+
+    // This would throw a NPE
```
Comment (dramaticlly) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:399]: is this comment still apply?

**Review by dramaticlly — APPROVED** [review]
thank you @Fokko for fixing this!

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 1 above)
Comment (Fokko) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java]: I agree that it is better to remove it, thanks!

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 2]
```diff
@@ -738,11 +738,16 @@ private static PartitionSpec freshSpec(int specId, Schema schema, PartitionSpec
...
+
+      final int fieldId;
+      if (sourceName != null) {
+        fieldId = schema.findField(sourceName).fieldId();
+      } else {
+        // In the case of a null sourceName, the column has been deleted.
+        // This only happens in V1 tables where the reference is still around as a void transform
+        fieldId = field.sourceId();
+      }
+      specBuilder.addField(field.transform().toString(), fieldId, field.fieldId(), field.name());
```
Comment (hililiwei) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:750]: nit: new empty line?

**Review by hililiwei — COMMENTED** [review]
LGTM, just one format comment.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 3]
```diff
@@ -386,4 +386,25 @@ public void testGroupingKeyTypeWithIncompatibleSpecEvolution() {
         "Conflicting partition fields",
         () -> Partitioning.groupingKeyType(table.schema(), table.specs().values()));
   }
+
+  @Test
+  public void testDeletingPartitionField() {
+    TestTables.TestTable table =
+        TestTables.create(tableDir, "test", SCHEMA, BY_DATA_SPEC, V1_FORMAT_VERSION);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:393]: should we maybe also add a test that does the same with `V2_FORMAT_VERSION`?

**Review by jackye1995 — APPROVED** [review]
looks good to me. I see Eduard has a comment about testing in v2, which I am not sure if is worth it, up to you!

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for fixing this, @Fokko!

**Pull request merged**